src/dbnode/storage/repair.go

```diff
@@ -299,6 +299,61 @@ func (r shardRepairer) Repair(
 	return metadataRes, nil
 }
 
+// TODO(rartoul): Currently throttling via the MemoryTracker can only occur at the level of an entire
+// block for a given namespace/shard/blockStart. For almost all practical use-cases this is fine, but
+// this could be improved and made more granular by breaking data that is being loaded into the shard
+// into smaller batches (less than one complete block). This would improve the granularity of throttling
+// for clusters where the number of shards is low.
+func (r shardRepairer) loadDataIntoShard(shard databaseShard, data result.ShardResult) error {
+	var (
+		waitingGauge  = r.scope.Gauge("waiting-for-limit")
+		waitedCounter = r.scope.Counter("waited-for-limit")
+		doneCh        = make(chan struct{})
+		waiting       bool
+		waitingLock   sync.Mutex
+	)
+	defer close(doneCh)
+
+	// Emit a gauge constantly that indicates whether or not the repair processed is blocking waiting.
+	go func() {
+		for {
+			select {
+			case <-doneCh:
+				waitingGauge.Update(0)
+				return
+			default:
+				waitingLock.Lock()
+				currWaiting := waiting
+				waitingLock.Unlock()
+				if currWaiting {
+					waitingGauge.Update(1)
+				} else {
+					waitingGauge.Update(0)
+				}
+				time.Sleep(5 * time.Second)
+			}
+		}
+	}()
+
+	for {
+		err := shard.Load(data.AllSeries())
+		if err == ErrDatabaseLoadLimitHit {
```

Comment:
Maybe this looks more idiomatic?
```go
var err error
for err = shard.Load(data.AllSeries()); err == ErrDatabaseLoadLimitHit {
	...
}

if err != nil {
	return err
}

return nil
```

Reply:
personally prefer the way I have it to not have it going so far to the right